From 8902daed95775febf11162e298083535cdb71081 Mon Sep 17 00:00:00 2001
From: Frederic Weisbecker <frederic@kernel.org>
Date: Sun, 16 Oct 2022 16:22:53 +0000
Subject: [PATCH 01/20] BACKPORT: FROMGIT: rcu: Fix missing nocb gp wake on
 rcu_barrier()

In preparation for RCU lazy changes, wake up the RCU nocb gp thread if
needed after an entrain.  This change prevents the RCU barrier callback
from waiting in the queue for several seconds before the lazy callbacks
in front of it are serviced.

Reported-by: Joel Fernandes (Google) <joel@joelfernandes.org>
Signed-off-by: Frederic Weisbecker <frederic@kernel.org>
Signed-off-by: Joel Fernandes (Google) <joel@joelfernandes.org>
Signed-off-by: Paul E. McKenney <paulmck@kernel.org>

(Backport: Adjustments due to changes in how wake_nocb_gp() locks)

(cherry picked from commit b8f7aca3f0e0e6223094ba2662bac90353674b04
 https://git.kernel.org/pub/scm/linux/kernel/git/paulmck/linux-rcu.git rcu/next)

Bug: 258241771
Bug: 222463781
Test: powerIdle lab tests.
Change-Id: I830269cd41b18862a1a58b26ce3292c6c4457bc7
Signed-off-by: Joel Fernandes <joelaf@google.com>
Reviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/3965070
Reviewed-by: Ross Zwisler <zwisler@google.com>
(cherry picked from commit 487156bf87b894bf5cd4ee886bf994b7bdc860d7)
[Cherry picked from chromeos-5.10 tree. Minor tweaks to commit message
to match Android style]
Signed-off-by: Qais Yousef <qyousef@google.com>

Upstream: b8f7aca3f0e0e6223094ba2662bac90353674b04 ("rcu: Fix missing nocb gp wake on rcu_barrier()")
ACK android13-5.10: fc4c0de2162ed3db506fac47a071e39c2ef0e859, taken from arter97/master
---
 kernel/rcu/tree.c | 17 +++++++++++++++--
 kernel/rcu/tree.h |  2 ++
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/kernel/rcu/tree.c b/kernel/rcu/tree.c
index f93568b1dc14..4f966a68922c 100644
--- a/kernel/rcu/tree.c
+++ b/kernel/rcu/tree.c
@@ -3857,12 +3857,22 @@ static void rcu_barrier_func(void *cpu_in)
 {
 	uintptr_t cpu = (uintptr_t)cpu_in;
 	struct rcu_data *rdp = per_cpu_ptr(&rcu_data, cpu);
+	bool wake_nocb = false;
+	bool was_alldone = false;
+	unsigned long flags;
 
 	rcu_barrier_trace(TPS("IRQ"), -1, rcu_state.barrier_sequence);
 	rdp->barrier_head.func = rcu_barrier_callback;
 	debug_rcu_head_queue(&rdp->barrier_head);
-	rcu_nocb_lock(rdp);
+	rcu_nocb_lock_irqsave(rdp, flags);
+	/*
+	 * Flush bypass and wakeup rcuog if we add callbacks to an empty regular
+	 * queue. This way we don't wait for bypass timer that can reach seconds
+	 * if it's fully lazy.
+	 */
+	was_alldone = rcu_segcblist_is_offloaded(&rdp->cblist) && !rcu_segcblist_pend_cbs(&rdp->cblist);
 	WARN_ON_ONCE(!rcu_nocb_flush_bypass(rdp, NULL, jiffies));
+	wake_nocb = was_alldone && rcu_segcblist_pend_cbs(&rdp->cblist);
 	if (rcu_segcblist_entrain(&rdp->cblist, &rdp->barrier_head)) {
 		atomic_inc(&rcu_state.barrier_cpu_count);
 	} else {
@@ -3870,7 +3880,10 @@ static void rcu_barrier_func(void *cpu_in)
 		rcu_barrier_trace(TPS("IRQNQ"), -1,
 				  rcu_state.barrier_sequence);
 	}
-	rcu_nocb_unlock(rdp);
+	if (wake_nocb)
+		wake_nocb_gp(rdp, false, flags);
+	else
+		rcu_nocb_unlock_irqrestore(rdp, flags);
 }
 
 /**
diff --git a/kernel/rcu/tree.h b/kernel/rcu/tree.h
index d67adf71541c..b1c556afde06 100644
--- a/kernel/rcu/tree.h
+++ b/kernel/rcu/tree.h
@@ -449,6 +449,7 @@ static void rcu_nocb_unlock_irqrestore(struct rcu_data *rdp,
 static void rcu_lockdep_assert_cblist_protected(struct rcu_data *rdp);
 #ifdef CONFIG_RCU_NOCB_CPU
 static void __init rcu_organize_nocb_kthreads(void);
+static void wake_nocb_gp(struct rcu_data *rdp, bool force, unsigned long flags);
 #define rcu_nocb_lock_irqsave(rdp, flags)				\
 do {									\
 	if (!rcu_segcblist_is_offloaded(&(rdp)->cblist))		\
@@ -458,6 +459,7 @@ do {									\
 } while (0)
 #else /* #ifdef CONFIG_RCU_NOCB_CPU */
 #define rcu_nocb_lock_irqsave(rdp, flags) local_irq_save(flags)
+static void wake_nocb_gp(struct rcu_data *rdp, bool force, unsigned long flags) {  }
 #endif /* #else #ifdef CONFIG_RCU_NOCB_CPU */
 
 static void rcu_bind_gp_kthread(void);
